syntax = "proto3";

package Protocol;
// import "google/protobuf/timestamp.proto"; 
option csharp_namespace = "Google.Protobuf.Protocol";

// 아래부터 공용으로 사용될 패킷용 구조체
enum ActorState {
    IDLE = 0;
    MOVING = 1;
	SKILL = 2;
    DEAD = 3;
}

enum ActorType {
	NONE = 0;
	PLAYER = 1;
	MONSTER = 2;
	PROJECTILE = 3; // 투사체, 화살같은 것
}

enum SkillType {
	SKILL_NONE = 0;
	SKILL_AUTO = 1; // 기본(자동)적으로 때리는 평타를 의미
	SKILL_PROJECTILE = 2;
}

enum MoveDir {
    Up = 0;
    Down = 1;
    Left = 2;
    Right = 3;
}

message ActorInfo {
	int32 actorId = 1;
	string name = 2;
	PositionInfo posInfo = 3;
	StatInfo statInfo = 4;
}

message PositionInfo {
	ActorState state = 1;
	MoveDir moveDir = 2;
	int32 posX = 3;
	int32 posY = 4;
}

message SkillInfo {
	int32 skillId = 1;
}

message StatInfo {
	int32 level = 1;
	int32 hp = 2;
	int32 maxHp = 3;
	int32 attack = 4;
	float speed = 5;
	int32 totalExp = 6;
}

// 아래부터 패킷 정보
// 패킷 형식
// S_ 서버 -> 클라이언트로 주는 패킷 prefix
// C_ 클라이언트 -> 서버로 주는 패킷 prefix
// 그 다음은 예시와 같은 형식으로 MsgId Enum 과 패킷 이름을 맞춰야 한다. ex) Enum S_ENTER_GAME 에 대응되는 패킷은 S_EnterGame 이 되어야 함. 

enum MsgId {
  S_ENTER_GAME = 0;
  S_LEAVE_GAME = 1;
  S_SPAWN = 2;
  S_DESPAWN = 3;
  C_MOVE = 4;
  S_MOVE = 5;
  C_SKILL = 6;
  S_SKILL = 7;
  S_CHANGE_HP = 8;
  S_DEAD = 9;
}

// 자신이 게임에 접속했음을 의미
message S_EnterGame {
  ActorInfo player = 1; // 플레이어에게만 전송되기 때문에 명칭을 player 로
}

// 자신이 게임에서 나갈 때를 의미
message S_LeaveGame {
}

// 타인이 게임에 이미 있거나 게임에 들어왔을 때
message S_Spawn {
	repeated ActorInfo actors = 1; // repeated 가 list 를 의미
}

// 타인이 게임에서 나갔을 떄
message S_Despawn {
	repeated int32 actorIds = 1;
}

message C_Move {
	PositionInfo posInfo = 1;
}

message S_Move {
	int32 actorId = 1;
	PositionInfo posInfo = 2;
}

message C_Skill {
	SkillInfo info = 1;
}

message S_Skill {
	int32 actorId = 1;
	SkillInfo info = 2;
}

message S_ChangeHp {
// TODO: 후에는 데미지 플로팅 같은 것을 위해서 더 세분화된 정보를 넣어야 한다. ex_ 사유, 상태
	int32 actorId = 1;
	int32 hp = 2; // 현재 체력 정보
}

message S_Dead {
	int32 actorId = 1; // 죽임당한 actor 의 id
	int32 attackerActorId = 2;
}